target/linux/ramips/image/mt7621.mk

```diff
@@ -136,6 +153,16 @@ define Device/hc5962
 endef
 TARGET_DEVICES += hc5962
 
+define Device/iodata_wn-ax1167gr
+  DTS := WN-AX1167GR
+  IMAGE_SIZE := 15552k
+  KERNEL_INITRAMFS := $$(KERNEL) | \
```

Comment:
WN-AX1167GR has dual os-image partition in stock firmware.

- ```dmesg```

```flash manufacture id: c2, device id 20 18
MX25L12805D(c2 2018c220) (16384 Kbytes)
mtd .name = raspi, .size = 0x01000000 (16M) .erasesize = 0x00010000 (64K) .numeraseregions = 0
Creating 10 MTD partitions on "raspi":
0x000000000000-0x000001000000 : "ALL"
0x000000000000-0x000000030000 : "Bootloader"
0x000000030000-0x000000040000 : "Config "
0x000000040000-0x000000050000 : "Factory"
0x000000050000-0x000000060000 : "iNIC_rf"
0x000000060000-0x0000007e0000 : "Kernel"
0x000000800000-0x000000f80000 : "app"
0x000000f90000-0x000000fa0000 : "Key"
0x000000fa0000-0x000000fb0000 : "backup"
0x000000fb0000-0x000001000000 : "storage"```

The os image is stored in "Kernel" and "app", and the value "7680k" is the size of those partitions and it is specified to fit them.

Also, I thought that I needed to create factory image manually using initramfs instead of sysupgrade_bin in this case.

Reply:
>Beside of this remark, the PR is as perfect as expected. Thanks a lot!

It is thanks to your precise advice. Thank you very much!

Reply:
> WN-AX1167GR has dual os-image partition in stock firmware.

I've to admit, I still don't get it, respectively the partition map of the original firmware suggest something different than you wrote.

First of all, I don't see an os-image partition in the dmesg output.

Do you mean by "dual os-image partition" that kernel+rootfs is stored in the kernel partition and a different set of kernel+rootfs is stored in the app partition? The stock firmware writes images either to the kernel partition XOR the app partition and sets a flag about the active partition. The bootloader evaluates this flag and reads the kernel either from the kernel partition XOR the app partition?

The equal size of the partitions suggests something like that. But how do you ensure that the "kernel" partition is the active one?  Following your installation instructions, it wouldn't work if the bootloader expects the "app" partition to be the active one and tries to read the kernel from this offset (0x7e0000) as we have our OpenWrt rootfs at this offset.

The naming of the stock firmware partitions rather suggests that the stock firmware upgrade code would write the kernel from our sysupgrade image to the Kernel partition and the rootfs from our image to the app partition. As OpenWrt expects the rootfs right behind the kernel, it would fail to boot (panic no rootfs found) in such a case.

I've some ideas to get it working without the extra initramfs step, but first I need to know how it really works.

Reply:
>Do you mean by "dual os-image partition" that kernel+rootfs is stored in the kernel partition and a different set of kernel+rootfs is stored in the app partition?

Yes. The flag for boot partition is stored in "Key" partition, and U-Boot reads this and determines the partition to boot.

In the stock firmware, each partition of "Kernel" and "app" consists of kernel + rootfs (ramfs) , and when updating the firmware, it writes the firmware to the other partition as a single image rather than the currently active partition.

If the second offset read by U-Boot is "Bad Magic Number", then it tries to boot from the first offset. Then, if the kernel is successfully read from the first offset, u-boot will change the flag and boot.

```
## Booting image at bc800000 ...
Bad Magic Number,FFFFFFFF
Boot from KERNEL 1  !!
## Booting image at bc060000 ...
   Image Name:	MIPS OpenWrt Linux-4.14.50
   Image Type:	MIPS Linux kernel Image (lzma compressed)
   Data Size:	1865917 Bytes = 1.8 MB
   Load Address:	80001000
   Entry Point:	80001000
   Verifying Checksum ... OK
   Uncompressing Kernel Image ... OK
raspi_erase_write: offs:f90000, count:34
.
.
Done!

Starting kernel ...
```

Reply:
> Yes. The flag for boot partition is stored in "Key" partition, and U-Boot reads this and determines the partition to boot.

Which means, in worst case the installation wouldn't work. If Kernel2 is active, it is impossible to boot, since the OpenWrt image is written to Kernel1 according to your installation instruction. Or do I miss here something? If I'm right, any idea about how to fix it? Maybe there is a command in u-boot to switch the active partition.

I'm still thinking about whether a factory image build the way I suggested in https://github.com/openwrt/openwrt/pull/1117#discussion_r198645924 would work. But most likely it will fail, if the stock firmware is started from Kernel1. The factory image is written to Kernel2 via the stock firmware, the OpenWrt kernel boots but starts looking for a rootfs beginning at the start of Kernel1. It either find a the rootfs not intended for OpenWrt or none at all and crashes/panics.

Reply:
this...

>If the second offset read by U-Boot is "Bad Magic Number", then it tries to boot from the first offset. Then, if the kernel is successfully read from the first offset, u-boot will change the flag and boot.

If booting from initramfs of the factory image in second partition (app) and perform sysupgrade and restarting ("app" partition is active), U-Boot tries to boot from the second partition at the first. But as it fails as you said, U-Boot will then automatically tries to boot from the first partition.

>Maybe there is a command in u-boot to switch the active partition.

Perhaps I think that this command is applicable.

```
MT7621 # help key
key [ .w, .r] vlaue
    - set key number
```

Reply:
Would you please add the information about

- the "Key" u-boot env
- the fact that stock OS uses a dual image layout 
- how the fallback works if the wrong active partition is selected (U-Boot is "Bad Magic Number")

to the commit message.

If something goes wrong and someone creates a bug report, the commit message would be my first stop to see how things should work/what kind of special handling is expected. As I most likely can't remember of the informations proved by your last comments if needed, would be great if we can document it in the commit message.